Enforce cursor be serializable

sambostock opened the pull request:
## Context & Problem

The `cursor_position` is serialized by the job adapter (Resque/Sidekiq) and bypasses Active Job's fancy serialization. Both Resque and Sidekiq effectively use `JSON.dump` to serialize it and `JSON.parse` to deserialize it.

In cases where a non-JSON-compatible cursor is used, it is silently turned into a `String` via `to_s`, which can lead to weird errors when the job resumes after interruption. These are not always straightforward to debug, and don't reliably show up in tests, as Active Support adds implicit coercions to some object's `==` method.

```ruby
time = Time.at(0).utc               # =>  1970-01-01 00:00:00 UTC  (Time)
time.to_s                           # => "1970-01-01 00:00:00 UTC" (String)
require 'json'
JSON.parse(JSON.dump(time))         # => "1970-01-01 00:00:00 UTC" (String)
time == JSON.parse(JSON.dump(time)) # => false                     (no coercion)
require 'active_support/all'
JSON.parse(JSON.dump(time))         # => "1970-01-01 00:00:00 UTC" (String)
time == JSON.parse(JSON.dump(time)) # => true                      (implicitly coerced)
```

<details><summary>Real world example</summary>
In one of our repos, we use a custom enumerator around a collection of events fetched from an API and ordered by time. We used the time the event occurred at as a cursor, which was "corrupted" by being serialized into a String. The API expected an ISO8601 timestamp (its wrapper gem converts `Time` objects automatically), but our improperly deserialized time strings were in the wrong format, so it blew up.

This went unnoticed for some time, due to the job not being interrupted, and was tricky to debug due to the reasons outlined above.
</details>

## What this PR does about it

**This prevents these errors by enforcing that the cursors be composed of basic Ruby objects.**

This is enforced by (recursively) analyzing the each cursor yielded by the enumerator **before `each_iteration`**. If it is composed of anything other than the following classes, a descriptive `CursorError` is raised:
- `Array`
- `Hash`
- `String`
- `Integer`
- `Float`
- `NilClass` (`nil`)
- `TrueClass` (`true`)
- `FalseClass` (`false`)

### Rationale

- It is better to verify each cursor and ensure we'd be able to interrupt before having started any work, than to finding out the cursor is unsafe after interruption.
- It is fastest and most reliable to simply analyze the component classes of the cursor (rather than check if serializing and deserializing yields the same object, due to implicit conversions).
- Since most cursors are simple Strings, Integers, or `nil`, the performance penalty of checking each cursor is negligible.

----
<details><summary>Original Description</summary>
The cursor ends up being serialized as a `String`, regardless of if it is a `String` or not (by calling `to_s`).

This can lead to subtle bugs often not found during testing.

For instance, if a `Time` is used as a cursor and the job is requeued, it instead receives the `Time#to_s` representation as its new input cursor.

Let's be explicit and save developers the headache of debugging that: **let's enforce that the cursor be an instance of `String` or `NilClass`.**
</details>

GustavoCaso commented on lib/job-iteration/iteration.rb:
```diff
@@ -251,5 +287,21 @@ def valid_cursor_parameter?(parameters)
+      else
+        object.instance_of?(String) ||
+          object.instance_of?(Integer) || object.instance_of?(Float) ||
+          object.equal?(nil) || object.equal?(true) || object.equal?(false)
```
We could probably dry the code a little here by doing:
```ruby
klasses = [String, Integer, Float, NilClass, TrueClass, FalseClass]
klasses.any? { |klass| object.instance_of?(klass) }
```

We might be even be able to extract `klasses` as a constant

sambostock replied:
It's funny, I actually went through having both those implementations but got caught up in trying to optimize for speed because this will run before `each_iteration`. 😅 

I figured we could unroll the loop, and use `.equal?` to quickly compare by object ID for singletons, instead of looking at their class. But benchmarking it shows a negligible difference, so I'll go with the constant array approach for readability. 👍

GustavoCaso commented on test/integration/sidekiq_test.rb:
```diff
@@ -39,6 +39,19 @@ class SidekiqIntegrationTest < ActiveSupport::TestCase
     assert_equal 0, queue_size
   end
 
+  test "Time cursor corruption is prevented" do
```
Could we add a test for Resque as well? Or does resque do not convert a Time cursor into a JSON?

sambostock replied:
I had only put this test in as a kind of "regression" test for the bug we experienced, but I can look at adding it to Resque's tests too. 👍

sambostock commented:
Thanks for the feedback @GustavoCaso! I have
- gone with the simple `CLASSES.any? { |klass| object.instance_of?(klass) }` approach
- added a Resque cursor corruption test

I also noticed the integration tests are basically identical for Resque and Sidekiq, so I extracted a common test module in #75, to keep it independent of this change.

rafaelfranca commented:
Should not Symbols be allowed?
